### PR TITLE
[ONNX] Add support for Trilu operator

### PR DESCRIPTION
**Pull request by mateusztabaka**
Ticket: 91096

### CODE REVIEW

File: src/frontends/onnx/frontend/src/op/trilu.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+    // fetch last two dimensions of input shape
+    // M = shape[-1]
+    // N = shape[-2]
+    const auto M = std::make_shared<default_opset::Gather>(shape,
+                                                           default_opset::Constant::create(element::i32, Shape{}, {-1}),
+                                                           zero);
+    const auto N = std::make_shared<default_opset::Gather>(shape,
+                                                           default_opset::Constant::create(element::i32, Shape{}, {-2}),
+                                                           zero);
```
Comment (rkazants) [inline src/frontends/onnx/frontend/src/op/trilu.cpp:69]: may it is better to use `Slice` operation for dimension retrieving because FWs use either `StridedSlice` or `Slice` and `Gather` behavior for negative indices is changing from one opset to the other.

File: src/frontends/onnx/frontend/src/ops_bridge.cpp [hunk 2]
```diff
@@ -241,10 +243,9 @@ OperatorSet OperatorsBridge::get_operator_set(const std::string& domain, int64_t
     }
     for (const auto& op : dm->second) {
         const auto& it = find(version, op.second);
-        if (it == std::end(op.second)) {
-            throw error::UnsupportedVersion{op.first, version, domain};
```
Comment (rkazants) [inline src/frontends/onnx/frontend/src/ops_bridge.cpp:245]: why is this removed?

File: src/bindings/python/tests/test_onnx/test_backend.py [hunk 3]
```diff
@@ -297,24 +297,8 @@ def expect_fail(test_case_path, xfail):  # type: (str) -> None
...
         "OnnxBackendNodeModelTest.test_tril_zero_cpu",
-        "OnnxBackendNodeModelTest.test_triu_cpu",
-        "OnnxBackendNodeModelTest.test_triu_neg_cpu",
-        "OnnxBackendNodeModelTest.test_triu_one_row_cpu",
-        "OnnxBackendNodeModelTest.test_triu_out_neg_out_cpu",
-        "OnnxBackendNodeModelTest.test_triu_out_pos_cpu",
-        "OnnxBackendNodeModelTest.test_triu_pos_cpu",
-        "OnnxBackendNodeModelTest.test_triu_square_cpu",
-        "OnnxBackendNodeModelTest.test_triu_square_neg_cpu",
         "OnnxBackendNodeModelTest.test_triu_zero_cpu",
```
Comment (rkazants) [inline src/bindings/python/tests/test_onnx/test_backend.py:302]: it looks duplicate of the previous line

**Review by rkazants — COMMENTED** [review]
In general, LGTM. Left some comments

File: src/bindings/python/tests/test_onnx/test_backend.py (hunk 3 above)
Comment (mateusztabaka) [reply src/bindings/python/tests/test_onnx/test_backend.py:302]: previous line refers to tril (case when upper == false) as opposed to current line (triu is a case when upper == true)

**Review by rkazants — APPROVED** [review]

File: src/bindings/python/tests/test_onnx/test_backend.py (hunk 3 above)
Comment (rkazants) [reply src/bindings/python/tests/test_onnx/test_backend.py:302]: OK

File: src/frontends/onnx/frontend/src/ops_bridge.cpp (hunk 2 above)
Comment (mateusztabaka) [reply src/frontends/onnx/frontend/src/ops_bridge.cpp:245]: I had a second look on this and actually I should have register Trilu to opset 1. So I can revert what I have removed.

The long story is that if I register Trilu to opset 14, the check in this loop breaks models that have opset version less than 14.
This loop scans every registered operator in ONNX frontend and if the operator version is greater than the version from the model that's being read - it throws this exception. For example, reading model with just Add node and opset version 7 will trigger this exception if any there's at least one operator in ONNX frontend that doesn't have version <= 7.

File: src/frontends/onnx/frontend/src/ops_bridge.cpp [hunk 4]
```diff
@@ -457,6 +458,7 @@ OperatorsBridge::OperatorsBridge() {
     REGISTER_OPERATOR("TopK", 10, topk);
     REGISTER_OPERATOR("TopK", 11, topk);
     REGISTER_OPERATOR("Transpose", 1, transpose);
+    REGISTER_OPERATOR("Trilu", 1, trilu);
```
Comment (rkazants) [inline src/frontends/onnx/frontend/src/ops_bridge.cpp:461]: Looks strange that we register this operation as from opset-1.
Not good workaround.

Comment (mitruska) [reply src/frontends/onnx/frontend/src/ops_bridge.cpp:461]: For now, this is and was a standard in approach for ONNX Frontend to register and make available new operations since onnx opset-1. Proper opset availability is crucial only for the next versions of the operation, if there are registered more than one. 
For example if ONNX introduce next Trilu version in onnx opset-18 we will have internally `set_1::Trilu` for onnx opset (1, 17 and `set_18::Trilu` for onnx opset (18, +) until the next version of the Trilu operation.

Of course we can discuss some improvements of this logic, but it is rather out of this PR scope.

**Review by mitruska — APPROVED** [review]

File: src/frontends/onnx/frontend/src/op/trilu.cpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    const auto inputs = node.get_ng_inputs();
+    const auto num_inputs = inputs.size();
+
+    CHECK_VALID_NODE(node, num_inputs > 0 && num_inputs <= 2, "Trilu expects <= 2 input tensors. Got: ", num_inputs);
+
+    const auto& input = inputs[0];
+    const auto& rank = input.get_partial_shape().rank();
+    if (rank.is_static()) {
+        CHECK_VALID_NODE(node, rank.get_length() >= 2, "Trilu first input's rank must be >= 2");
+    }
```
[the remark this answers is not in the document]
Comment (mateusztabaka) [reply src/frontends/onnx/frontend/src/op/trilu.cpp:26]: not really, rank can be greater or equal 2

[the remark this answers is not in the document]
Comment (mitruska) [reply src/frontends/onnx/frontend/src/op/trilu.cpp:26]: Oh my bad, sure then probably something like `rank.is_compatible(Dimension(2, -1))` could work, but let's leave it as is:)

File: src/frontends/onnx/frontend/src/ops_bridge.cpp (hunk 4 above)
Comment (mateusztabaka) [reply src/frontends/onnx/frontend/src/ops_bridge.cpp:461]: Registering to new ops to opset1 is kind of legacy approach we still follow. However, technically (except this UnsupportedVersion exception) I can't see any reason we couldn't register Trilu starting from opset 14 instead of opset 1, regardless if we could have next version of Trilu in e.g. opset 18 or not. But in my opinion, we can think about that later and keep this PR as is.

**Review by rkazants — APPROVED** [review]

File: src/bindings/python/tests_compatibility/test_onnx/test_backend.py [hunk 6]
```diff
@@ -296,26 +297,14 @@ def expect_fail(test_case_path, xfail):  # type: (str) -> None
...
-        "OnnxBackendNodeModelTest.test_triu_out_neg_out_cpu",
-        "OnnxBackendNodeModelTest.test_triu_out_pos_cpu",
-        "OnnxBackendNodeModelTest.test_triu_pos_cpu",
-        "OnnxBackendNodeModelTest.test_triu_square_cpu",
-        "OnnxBackendNodeModelTest.test_triu_square_neg_cpu",
         "OnnxBackendNodeModelTest.test_triu_zero_cpu",
     ),
+    (
+        skip_dynamic_model,
+        "OnnxBackendNodeModelTest.test_triu_one_row_cpu"
```
[the remark this answers is not in the document]
Comment (mateusztabaka) [reply src/bindings/python/tests_compatibility/test_onnx/test_backend.py:306]: it's kind of lucky case. Select output is dynamic:
Select({?,5}, {3, 1, 5}, {}) -> {3,?,5}
because of Select's second input has a unit dimension

File: src/frontends/onnx/frontend/src/op/trilu.cpp [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    bool is_k_available = num_inputs == 2 && !ngraph::op::is_null(inputs[1]);
+    if (is_k_available) {
+        CHECK_VALID_NODE(node, inputs[1].get_partial_shape().compatible({}), "Trilu second input must be a scalar");
+    }
+
+    const auto shape = std::make_shared<default_opset::ShapeOf>(input);
```
Comment (jane-intel) [inline src/frontends/onnx/frontend/src/op/trilu.cpp:32]: I would suggest to explicitly state output type of ShapeOf as
1. k input of Trilu is int64
2. we do the same for Range operator since both Range and ShapeOf has additional input for output_type

`const auto shape = std::make_shared<default_opset::ShapeOf>(input, element::i64)`
